### PR TITLE
Migrate `snipes` and `snipesFor` in tests to `cleanByImpersonation` or `marketOrder*`

### PR DESCRIPTION
**Pull request by espendk**
This PR completes the replacement of `snipes` with `cleanByImpersonation` started in  #437 by migrating all tests away from `snipes` and `snipesFor`. 

TODO:
- [ ]  The main test in `Scenarii.t.sol` relied on snipe. It is not clear to me whether it's a test of sniping (and thus can be deleted) or if it also tests something else (and thus should be removed)? @jkrivine maybe knows?

### CODE REVIEW

File: test/core/Gatekeeping.t.sol
Comment (adhusson) [inline test/core/Gatekeeping.t.sol:322]: Is there a similar test for clean's max volume?

File: test/core/Gatekeeping.t.sol
Comment (adhusson) [inline test/core/Gatekeeping.t.sol:337]: Is there a test that checks clean cannot decrease the `taker` balance?

File: test/core/Gatekeeping.t.sol
Comment (adhusson) [inline test/core/Gatekeeping.t.sol:445]: Here and in the other tests where `x_on_y_succeeds|fails` I'm not seeing a check that the callbacks were indeed called. There wasn't a check before but it was very likely since the offer with the callback was being sniped. Here if there is a change to `setUp` in this contract or the parent contracts and a cheap offer is posted to the market and succeeds, then won't the tests become vacuously true?

File: test/core/MakerOperations.t.sol
Comment (adhusson) [inline test/core/MakerOperations.t.sol]: Similarly in this file I this it is less clear that the callback are being called. I think something like a `callbackWasCalled()` on each test would be enough, and `callbackWasCalled` checks that some state variable has been flipped from false to true?

edit: I see we do it in `MakerPosthook.t.sol`!

File: test/core/MakerPosthook.t.sol
Comment (adhusson) [inline test/core/MakerPosthook.t.sol:424]: Except here -- we don't check the posthook was called

File: test/core/Scenarii.t.sol
Comment (adhusson) [inline test/core/Scenarii.t.sol:272]: For this I think we need the advice of @jkrivine

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol]: The purpose of this test is to make sure inbound volumes are rounded up when partially taking an offer i.e. you can't have the taker pay 0 if the maker sends > 0 to the taker.

The test sets this up with a wants=9, gives=10 offer,and the taker asks for a volume of 1.

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol:121]: Is there a `clean` version of this test that should exist somewhere?

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol:211]: Similarly, shouldn't we test something similar with clean (ie that cleaning fails in this setup)?

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol]: I don't understand this comment

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol]: Same as above!

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol:658]: Do we have tests for clean on that topic?

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol]: Ideally, we'd include the credited value in the emitted event. However, the penalty isn't something we can calculate (since it depends on actual gas usage) and I think it's likely to change over time, so putting in a constant would require maintenance.

Do you think we should something different?

File: test/core/Gatekeeping.t.sol
Comment (espendk) [reply test/core/Gatekeeping.t.sol:322]: Yes, in `TakerOperations.t.sol` there's `test_gives_volume_above_96bits_fails_clean` and `test_gives_tick_outside_range_fails_clean`.

File: test/core/Gatekeeping.t.sol
Comment (espendk) [reply test/core/Gatekeeping.t.sol:337]: In  `TakerOperations.t.sol` there's `test_cleaning_by_impersonation_succeeds_and_does_not_transfer_funds`.

Do you think we need more/smth different?

File: test/core/Gatekeeping.t.sol
Comment (espendk) [reply test/core/Gatekeeping.t.sol:445]: The assertion two lines down (L448) is checking that the gasreq of the offer posted in line 442 has been changed as defined in the posthook. Both numbers are defined in the test, are unaffected by `setUp`, and (I believe) can only be changed by the posthook (the offer id is passed from the test). So it should be ok?

Comment (espendk) [reply test/core/Gatekeeping.t.sol:445]: Thinking some more on this: I think was we need is an assertion that the expected offer was executed since it's not guaranteed by `marketOrder` whereas it was with `snipe`. The tests in `Posthook.t.sol` should allow us to assume that the posthook is called if the offer is executed.

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol:121]: Yes, `test_clean_multiple_failing_offers` in the same file.

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol:211]: The test `test_cleaning_non_failing_offer_fails` tests the same, just in a slightly different way (it was written before migrating the snipes tests).

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol]: Same answer/question as above :-)

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol:658]: Yes, `test_giving_smaller_gasreq_to_clean_fails` in the same file.

Comment (adhusson) [reply test/core/TakerOperations.t.sol:658]: Sorry I should have checked better that all those tests I asked about had cleaning counterparts!

Comment (espendk) [reply test/core/TakerOperations.t.sol:658]: np :-)

File: test/core/Gatekeeping.t.sol
Comment (espendk) [reply test/core/Gatekeeping.t.sol:445]: I'm implementing this and it's cleaner + makes things easier to read

Comment (espendk) [reply test/core/Gatekeeping.t.sol:445]: Addressed in [6d1388d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/6d1388d7700b735d4d362afbe231ab09109b62bd)

File: test/core/MakerOperations.t.sol
Comment (espendk) [reply test/core/MakerOperations.t.sol]: addressed in [6d1388d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/6d1388d7700b735d4d362afbe231ab09109b62bd)

File: test/core/MakerPosthook.t.sol
Comment (espendk) [reply test/core/MakerPosthook.t.sol:424]: Not sure how we can check it, since the posthook reverts and thus we cannot retain any state change made in it?

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol]: I found this pattern in `MakerPosthook.t.sol`, I'll copy that instead:

```
    //TODO: when events can be checked instead of expected, take given penalty instead of ignoring it
    vm.expectEmit(true, true, true, false, $(mgv));
```

Comment (espendk) [reply test/core/TakerOperations.t.sol]: Addressed in [a21d89d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/a21d89dbc0f556eda8cdea09e69e21f89da9b907)

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol]: Addressed in [a21d89d](https://github.com/mangrovedao/mangrove-core/pull/454/commits/a21d89dbc0f556eda8cdea09e69e21f89da9b907)

File: test/core/TakerOperations.t.sol
Comment (espendk) [reply test/core/TakerOperations.t.sol]: Ok, thanks for the explanation :-) I've copied it into a comment, see [8f2c2cb](https://github.com/mangrovedao/mangrove-core/pull/454/commits/8f2c2cb6413e55228132de90f317cde596c0fde1)

File: test/core/MakerPosthook.t.sol
Comment (adhusson) [reply test/core/MakerPosthook.t.sol:424]: I tried to check Mangrove's corresponding log: 3994d96ff8b77f2486f322254aa3c1860b4ad9a5

**Review by adhusson — APPROVED** [review]
pending my small test-editing commit lgtm

**Comment by espendk** [comment]
> pending my small test-editing commit lgtm

Perfect, thanks.

I'll merge once @jkrivine has had a chance to weigh in re. Scenarii.t.sol.

**Comment by espendk** [comment]
I merged the PR with the FIXME re. Scenarii in as @jkrivine  is busy and we need to move on :-)
We'll address the Scenarii test later - I'll add a task on it to the ticks PR.

**Pull request merged**
